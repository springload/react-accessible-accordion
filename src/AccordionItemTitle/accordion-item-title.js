// @flow

import React, { Component } from 'react';
import type { Node } from 'react';
import classNames from 'classnames';
import { inject, observer } from 'mobx-react';

type AccordionItemTitleProps = {
    children: Node,
    className: string,
    hideBodyClassName: string,
    accordionStore: {
        items: Array<Object>,
        accordion: boolean,
        onChange: Function,
    },
    itemkey: string | number,
};

type AccordionItemTitleState = {};

export class AccordionItemTitle extends Component<
    AccordionItemTitleProps,
    AccordionItemTitleState,
> {
    static accordionElementName = 'AccordionItemTitle';

    static defaultProps = {
        className: 'accordion__title',
        hideBodyClassName: '',
    };

    handleClick() {
        const { itemkey } = this.props;
        const { accordion, onChange } = this.props.accordionStore;
        let { items } = this.props.accordionStore;
        const foundItem = items.find(item => item.itemkey === itemkey);
        if (!foundItem) return;

        if (accordion) {
            const newValue = !foundItem.expanded;
            items = items.map(item => {
                const resetItem = item;
                resetItem.expanded = false;
                return resetItem;
            });
            foundItem.expanded = newValue;
        } else {
            foundItem.expanded = !foundItem.expanded;
        }

        if (accordion) {
            onChange(foundItem.itemkey);
        } else {
            const filteredItems = items.filter(item => item.expanded === true);
            const itemkeys = filteredItems.map(item => item.itemkey);
            onChange(itemkeys);
        }
    }

    handleClick = this.handleClick.bind(this);

    handleKeyPress(evt: SyntheticInputEvent<HTMLButtonElement>) {
        if (evt.charCode === 13 || evt.charCode === 32) {
            this.handleClick();
        }
    }

    handleKeyPress = this.handleKeyPress.bind(this);

    render() {
        const { items, accordion } = this.props.accordionStore;
        const { itemkey, children, className, hideBodyClassName } = this.props;
        const foundItem = items.find(item => item.itemkey === itemkey);
        if (!foundItem) return null;

        const { itemUuid, expanded } = foundItem;

        const id = `accordion__title-${itemUuid}`;
        const ariaControls = `accordion__body-${itemUuid}`;
        const role = accordion ? 'tab' : 'button';
        const titleClassName = classNames(className, {
            [hideBodyClassName]: hideBodyClassName && !expanded,
        });

        if (role === 'tab') {
            return (
                <div
                    id={id}
                    aria-selected={expanded}
                    aria-controls={ariaControls}
                    className={titleClassName}
                    onClick={this.handleClick}
                    role={role}
                    tabIndex="0" // eslint-disable-line jsx-a11y/no-noninteractive-tabindex
                    onKeyPress={this.handleKeyPress}
                >
                    {children}
                </div>
            );
        }
        return (
            <div
                id={id}
                aria-expanded={expanded}
                aria-controls={ariaControls}
                className={titleClassName}
                onClick={this.handleClick}
                role={role}
                tabIndex="0" // eslint-disable-line jsx-a11y/no-noninteractive-tabindex
                onKeyPress={this.handleKeyPress}
            >
                {children}
            </div>
        );
    }
}

export default inject('accordionStore')(observer(AccordionItemTitle));
