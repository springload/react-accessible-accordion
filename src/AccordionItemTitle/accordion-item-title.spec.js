// @flow

import React from 'react';
import renderer from 'react-test-renderer';
import { AccordionItemTitle } from './accordion-item-title';

describe('AccordionItemTitle', () => {
    let mockAccordionStore;

    beforeEach(() => {
        mockAccordionStore = {
            items: [{
                itemkey: 'item-one-itemkey',
                expanded: false,
                itemUuid: 'item-one-itemUUid',
            },
            {
                itemkey: 'item-two-itemkey',
                expanded: true,
                itemUuid: 'item-two-itemUUid',
            }],
            accordion: false,
            onChange: jest.fn(),
        };
    });

    it('renders correctly with min params', () => {
        const tree = renderer.create(
            <AccordionItemTitle accordionStore={mockAccordionStore} itemkey="item-one-itemkey">
                <div>Fake Title</div>
            </AccordionItemTitle>,
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('renders correctly with different className', () => {
        const tree = renderer.create(
            <AccordionItemTitle className="testCSSClass" accordionStore={mockAccordionStore} itemkey="item-one-itemkey">
                <div>Fake Title</div>
            </AccordionItemTitle>,
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('renders with different hideBodyClassName', () => {
        const tree = renderer.create(
            <AccordionItemTitle hideBodyClassName="testCSSClass--hidden" accordionStore={mockAccordionStore} itemkey="item-one-itemkey">
                <div>Fake title</div>
            </AccordionItemTitle>,
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('doesn\'t respect hideBodyClassName when collapsed', () => {
        const tree = renderer.create(
            <AccordionItemTitle hideBodyClassName="testCSSClass--hidden" accordionStore={mockAccordionStore} itemkey="item-two-itemkey">
                <div>Fake title</div>
            </AccordionItemTitle>,
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('renders correctly when pressing enter', async () => {
        const tree = renderer.create(
            <AccordionItemTitle accordionStore={mockAccordionStore} itemkey="item-one-itemkey">
                <div>Fake Title</div>
            </AccordionItemTitle>,
        );

        tree.getInstance().handleKeyPress({ charCode: 13 });

        await new Promise(resolve => setTimeout(resolve, 200));

        expect(mockAccordionStore.onChange).toHaveBeenCalledTimes(1);
        expect(tree).toMatchSnapshot();
    });

    it('renders correctly when pressing space', async () => {
        const tree = renderer.create(
            <AccordionItemTitle accordionStore={mockAccordionStore} itemkey="item-one-itemkey">
                <div>Fake Title</div>
            </AccordionItemTitle>,
        );

        tree.getInstance().handleKeyPress({ charCode: 32 });

        await new Promise(resolve => setTimeout(resolve, 200));

        expect(mockAccordionStore.onChange).toHaveBeenCalledTimes(1);
        expect(tree).toMatchSnapshot();
    });

    it('renders correctly when pressing another key', async () => {
        const tree = renderer.create(
            <AccordionItemTitle accordionStore={mockAccordionStore} itemkey="item-one-itemkey">
                <div>Fake Title</div>
            </AccordionItemTitle>,
        );

        tree.getInstance().handleKeyPress({ charCode: 35 });

        await new Promise(resolve => setTimeout(resolve, 200));

        expect(mockAccordionStore.onChange).toHaveBeenCalledTimes(0);
        expect(tree).toMatchSnapshot();
    });
});
