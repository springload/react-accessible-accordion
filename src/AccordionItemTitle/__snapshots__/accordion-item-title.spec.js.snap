// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccordionItemTitle doesn't respect hideBodyClassName when collapsed 1`] = `
<div
  aria-controls="accordion__body-item-two-itemUUid"
  aria-expanded={true}
  className="accordion__title"
  id="accordion__title-item-two-itemUUid"
  onClick={[Function]}
  onKeyPress={[Function]}
  role="button"
  tabIndex="0"
>
  <div>
    Fake title
  </div>
</div>
`;

exports[`AccordionItemTitle renders correctly when pressing another key 1`] = `
<div
  aria-controls="accordion__body-item-one-itemUUid"
  aria-expanded={false}
  className="accordion__title"
  id="accordion__title-item-one-itemUUid"
  onClick={[Function]}
  onKeyPress={[Function]}
  role="button"
  tabIndex="0"
>
  <div>
    Fake Title
  </div>
</div>
`;

exports[`AccordionItemTitle renders correctly when pressing enter 1`] = `
<div
  aria-controls="accordion__body-item-one-itemUUid"
  aria-expanded={false}
  className="accordion__title"
  id="accordion__title-item-one-itemUUid"
  onClick={[Function]}
  onKeyPress={[Function]}
  role="button"
  tabIndex="0"
>
  <div>
    Fake Title
  </div>
</div>
`;

exports[`AccordionItemTitle renders correctly when pressing space 1`] = `
<div
  aria-controls="accordion__body-item-one-itemUUid"
  aria-expanded={false}
  className="accordion__title"
  id="accordion__title-item-one-itemUUid"
  onClick={[Function]}
  onKeyPress={[Function]}
  role="button"
  tabIndex="0"
>
  <div>
    Fake Title
  </div>
</div>
`;

exports[`AccordionItemTitle renders correctly with different className 1`] = `
<div
  aria-controls="accordion__body-item-one-itemUUid"
  aria-expanded={false}
  className="testCSSClass"
  id="accordion__title-item-one-itemUUid"
  onClick={[Function]}
  onKeyPress={[Function]}
  role="button"
  tabIndex="0"
>
  <div>
    Fake Title
  </div>
</div>
`;

exports[`AccordionItemTitle renders correctly with min params 1`] = `
<div
  aria-controls="accordion__body-item-one-itemUUid"
  aria-expanded={false}
  className="accordion__title"
  id="accordion__title-item-one-itemUUid"
  onClick={[Function]}
  onKeyPress={[Function]}
  role="button"
  tabIndex="0"
>
  <div>
    Fake Title
  </div>
</div>
`;

exports[`AccordionItemTitle renders with different hideBodyClassName 1`] = `
<div
  aria-controls="accordion__body-item-one-itemUUid"
  aria-expanded={false}
  className="accordion__title testCSSClass--hidden"
  id="accordion__title-item-one-itemUUid"
  onClick={[Function]}
  onKeyPress={[Function]}
  role="button"
  tabIndex="0"
>
  <div>
    Fake title
  </div>
</div>
`;
